torch/lib/c10d/Utils.hpp

```diff
@@ -300,8 +300,10 @@ inline at::Tensor newLikeFlat(
   }
   at::DeviceGuard gpuGuard(device);
   std::vector<int64_t> sizes{static_cast<int64_t>(tensors[deviceIdx].size())};
+  std::vector<int64_t> strides{static_cast<int64_t>(t.numel())};
   sizes.insert(sizes.end(), t.sizes().begin(), t.sizes().end());
-  return at::empty(sizes, t.options());
+  strides.insert(strides.end(), t.strides().begin(), t.strides().end());
+  return at::empty_strided(sizes, strides, t.options().memory_format(c10::nullopt));
```

Comment:
I am a bit confused about what this function is doing.
Also why this one is modified but not the overload just below?

Reply:
It's creating a flat buffer whose shape is the concatenation of all the (identically sized) individual tensors that are being gathered or scattered.  Say we are allgathering tensors from 4 devices (for example, 2 processes with 2 devices each, or 4 processes with 1 device each).  On every device, the source tensor `src` has size `(sz0, sz1, sz2)` and strides `(st0, st1, st2)` (elsewhere, we assert `src` is non overlapping and dense).  On every device, the buffer to hold the allgathered result is created as above, an empty tensor with size `(4, sz0, sz1, sz2)` and strides `(src.numel(), st0, st1, st2)`.  Then on each process we pass the raw `src` pointer and raw empty data pointer to the nccl allgather, and the nccl allgather fills the empty tensor with data pulled from the raw memory of the `src` tensors on each process.  Now it's easy to move data back out into the final destination tensors.  In the allgather epilogue. that looks like
```
outputTensors[i][j].copy_(outputFlattened[i][j], true);
```
Here `i` is the index over devices within this process (accounting for the possibility that a single process handles multiple GPUs) and `j` is the index over user-supplied output tensors (of which there are 4, one to receive the result from each device in the comm's world). `outputFlattened[i][0]`, for example, is the first dense slice of the (formerly empty) output buffer that contains the data pulled in from the first device in the comm's world, and it views the slice with sizes `(sz0, sz1, sz2)` and strides `(st0, st1, st2)`, which is correct because it matches the raw layout nccl pulled in from that `src` tensor.  So the `copy_` to `outputTensors[i][j]` will be semantically correct no matter what strides the output tensor has.

To be safe, I still think it's important for me to add a note to the distributed docs saying that all individual tensor inputs to a given collective must have the same sizes and strides both within each call in each process as well as across processes.  (e.g. if the `src` tensors have different layouts on different processes, the logic breaks.  But the new functionality is less restrictive than current code, for which all inputs must be rowmajor contiguous, period.)

> Also why this one is modified but not the overload just below?

Turns out ProcessGroupNccl doesn't need or use the overload below.

Reply:
Who uses the overload below then? Will your changes break all the distributed backends except nccl with nhwc? It's probably not a big deal, but should be mentioned.

Reply:
My changes shouldn't break other backends, because they should still assert their inputs and outputs are rowmajor contiguous, which newLikeFlat still supports (it must).  What is your worry here?  Other backends don't do strict enough contiguity checks and by making newLikeFlat more flexible i somehow give them more rope to hang themselves?

ProcessGroupNccl is the only place I relaxed the stride check to is_nonoverlapping_and_dense(), so Nccl will be the only backend that's known to support channels last (or otherwise permuted) dense tensors.  If I make that clear in the docs, is it acceptable?

Reply:
My worry is that non-contiguous gradients are going to be produced now, in particular for channels-last runs, and if someone is doing distributed channels-last run with a backend other than nccl, it's going to break (whereas it was working before). I still think it is acceptable, because a fraction of people using distributed channels-last without nccl backend is about 0? @albanD what's your take on this?

Reply:
What about people doing distributed CPU-only training? We definitely have users doing that internally. But I'm not sure which backend they use exactly.

Is the idea of not modifying the overload below because it is not used by NCCL but might be by other backends?

Reply:
I think DDP with non-nccl backends will be alright.  DDP's flat allreduce tensors remain rowmajor contiguous in this PR (I only altered the shapes of some views into them) so should still work with any backend.  Modifications to the raw nccl comms aren't needed to support DDP in particular, but like I said in the description
> The primitive comm changes aren't necessary to enable the DDP changes, but I wasn't sure this would end up true until I had written both sets of changes.

With non-contiguous gradients, the case that could break is if people are doing distributed non-nccl training with manual raw (as opposed to DDP-orchestrated) comms.